build-system/compile/json-schema/index.js

```diff
@@ -0,0 +1,221 @@
+    },
+  })),
+};
+
+const getAjv = once(() => new Ajv(ajvOptions));
+
+const escapeJsIdentifier = (id) => id.replace(/[^a-z_0-9]/gi, '_');
+
+const getTransformCache = once(() => new TransformCache('.json-schema-cache'));
+
+/**
+ * Like batchedRead(), but also includes `ajvOptions` in the invalidation hash,
+ * including the definition of functions.
+ * @param {string} filename
+ * @return {Promise<{contents: string, hash: string}>}
+ */
+async function ajvBatchedRead(filename) {
+  const {contents, hash} = await batchedRead(filename);
+  const hashable = {ajvOptions, hash};
+  const rehash = md5(jsonStringifyFunctions(hashable));
+  return {contents, hash: rehash};
+}
+
+const jsonStringifyFunctions = (value) =>
+  JSON.stringify(value, (_, value) =>
+    typeof value === 'function' ? value.toString() : value
+  );
+
+/**
+ * @param {string} filename
+ * @return {Promise<string>}
+ */
+async function getCompiledJsonSchemaFilename(filename) {
+  const {contents, hash} = await ajvBatchedRead(filename);
+  const transformCache = getTransformCache();
+  const cached = await transformCache.get(hash);
+  let output = cached;
+  if (!output) {
+    const schema = json5.parse(contents);
+    output = avjCompile(filename, schema);
+    transformCache.set(hash, Promise.resolve(output));
+  }
+  const outputFilename = `build/${filename}.js`;
+  if (!cached || !(await pathExists(outputFilename))) {
+    await outputFile(outputFilename, output);
+  }
+  return outputFilename;
+}
+
+/**
+ * @param {string} filename
+ * @param {any} schema
+ * @return {string}
+ */
+function avjCompile(filename, schema) {
+  const ajv = getAjv();
+  const validateAjv = ajv.compile(addCustomValidatorsToSchema(schema));
+  const scopeCode = ajv.scope.scopeCode(validateAjv?.source?.scopeValues, {});
+  const validateFnName = escapeJsIdentifier(
+    `validate_${basename(filename, '.json')}`
+  );
+  const customValidatorFns = customValidators.map(({fn}) => fn).join(', ');
+  return (
+    dedent(`
+      import {__customValidatorFns__} from '__customValidatorsModule__';
+
+      __scopeCode__
+
+      const validateAjv = __validateAjv__;
+
+      /**
+       * @param {any} data
+       * @return {{params?: object, message?: string, data?: any}[]}
+       * https://ajv.js.org/api.html#error-objects
+       */
+      export default function __validateFnName__(data) {
+        return validateAjv(data) ? [] : [...validateAjv.errors]
+      }
+    `)
+      .replace('__customValidatorFns__', customValidatorFns)
+      .replace('__customValidatorsModule__', customValidatorsModule)
+      .replace('__scopeCode__', scopeCode)
+      .replace('__validateFnName__', validateFnName)
+      // validateAjv returns a boolean, and modifies a property of the function
+      // for errors. The default instead returns an array of errors, which is
+      // empty when the input is valid.
+      .replace('__validateAjv__', validateAjv?.source?.validateCode)
```

Comment:
Continuing https://github.com/ampproject/amphtml/pull/37794/#discussion_r817182457:

I don't understand what wrapping is going on.  It seems we want to export a `__validateFnName__` so that we can return an array of `0...N` errors? But that doesn't require a wrapper, just a bit of string to append. It seems like we'd be able to remove internal AJV details by not using `validateCode`.

Reply:
To my knowledge, `validateCode` is the lowest level at which we can get the compiled code, so there's no way to manipulate the return value inside the function.

Alternatively, we can transform the output to get what we want. This would be safe, and would help us reduce size.

```diff
- let vErrors = null;
+ const vErrors = [];
  let errors = 0;
  if (!isValidCurrencyCode(data)) {
    const err0 = { ... };
-   if (vErrors === null) {
-     vErrors = [err0];
-   } else {
      vErrors.push(err0);
-   }
    errors++;
  }
- validate1.errors = vErrors;
- return errors === 0;
+ return vErrors;
```

Reply:
`validateCode` is an undocumented feature. I'm just asking you to switch to the [standalone](https://ajv.js.org/standalone.html) module:

```diff
diff --git a/build-system/compile/json-schema/index.js b/build-system/compile/json-schema/index.js
index 5e1f79f5da..2d307352af 100644
--- a/build-system/compile/json-schema/index.js
+++ b/build-system/compile/json-schema/index.js
@@ -7,6 +7,7 @@
  *   ```
  */
 const {_: ajvCode, Name: AjvName, default: Ajv} = require('ajv');
+const {default: AjvStandalone} = require('ajv/dist/standalone');
 const dedent = require('dedent');
 const json5 = require('json5');
 const {outputFile, pathExists} = require('fs-extra');
@@ -168,7 +169,7 @@ async function getCompiledJsonSchemaFilename(filename) {
 function avjCompile(filename, schema) {
   const ajv = getAjv();
   const validateAjv = ajv.compile(addCustomValidatorsToSchema(schema));
-  const scopeCode = ajv.scope.scopeCode(validateAjv?.source?.scopeValues, {});
+  const code = AjvStandalone(ajv, validateAjv);
   const validateFnName = escapeJsIdentifier(
     `validate_${basename(filename, '.json')}`
   );
@@ -177,8 +178,6 @@ function avjCompile(filename, schema) {
     dedent(`
       import {__customValidatorFns__} from '__customValidatorsModule__';
 
-      __scopeCode__
-
       __validateAjv__
 
       /**
@@ -192,12 +191,11 @@ function avjCompile(filename, schema) {
     `)
       .replace('__customValidatorFns__', customValidatorFns)
       .replace('__customValidatorsModule__', customValidatorsModule)
-      .replace('__scopeCode__', scopeCode)
       .replace('__validateFnName__', validateFnName)
       // validateAjv returns a boolean, and modifies a property of the function
       // for errors. The default instead returns an array of errors, which is
       // empty when the input is valid.
-      .replace('__validateAjv__', validateAjv?.source?.validateCode)
+      .replace('__validateAjv__', code)
       .replace(/__validateAjvName__/g, validateAjv?.source?.validateName.str)
   );
 }
```

Reply:
That would be fine, with some caveats:

- We still need the wrapper function to change the signature. 
- `AjvStandalone` adds `export` statements. We need to transform the generated code to remove them.